File: 백준/Gold/1774. 우주신과의 교감/우주신과의 교감.py
#Authored by : chjh2129
import sys, heapq
'''
  8 2
  1 2
  5 5
  2 1 
  9 9
  24 21
  6 -1
  0 0
  8 8
  1 2
  3 4

  다음과 같은 예제에서 처리를 잘해야합니다.

  미리 연결되어 있는 정점들은 가중치가 0인 간선을 추가로 인접리스트에 추가한 다음

  프림 알고리즘을 사용하면 풀이할 수 있습니다.
'''
input = sys.stdin.readline

def main() -> None:
  # input & init
  N, M = map(int, input().split())
  
  G: list[list[tuple]] = [ [] for _ in range(N+1) ]  
  chk: list[bool] = [ False for _ in range(N+1) ]
  
  Pos: list[tuple] = [(0, 0)] + [ (x, y) for x, y in (map(int, input().split()) for _ in range(N)) ]

  for u, v in (map(int, input().split()) for _ in range(M)):
    G[u].append((0, v))
    G[v].append((0, u))

  # solve
  for i in range(1, N):
    for j in range(i+1, N+1):
      dx: int = Pos[i][0] - Pos[j][0]
      dy: int = Pos[i][1] - Pos[j][1]
      dist: int = dx*dx + dy*dy
      G[i].append((dist, j))
      G[j].append((dist, i))
  
  pq: heapq[tuple] = []
  chk[1] = True
  for cost, nxt in G[1]: heapq.heappush(pq, (cost, 1, nxt))

  cnt: int = 0
  ans: float = 0.0

  while cnt < N - 1:
    dist, prv, cur = heapq.heappop(pq)
    if not chk[cur]:
      chk[cur] = True
      cnt += 1
      ans += dist ** 0.5

      for nxdist, nxt in G[cur]:
        if not chk[nxt]:
          heapq.heappush(pq, (nxdist, cur, nxt))
  
  # output
  print("%.2f" % ans)
  
sys.exit(main())

File: 백준/Gold/1774. 우주신과의 교감/우주신과의 교감.java
import java.io.*;
import java.util.*;
/*
  미리 연결되어 있는 정점들은 가중치가 0인 간선을 추가로 인접리스트에 추가한 다음

  프림 알고리즘을 사용하면 풀이할 수 있습니다. 
*/
public class Main {
  
  public static int n, m;
  public static int[][] pos;
  public static ArrayList<ArrayList<long[]>> adj = new ArrayList<>();
  public static boolean[] chk;
  public static PriorityQueue<long[]> pq = new PriorityQueue<>(Comparator.comparingLong(e -> e[0]));

  public static void main(String[] args) throws IOException {
    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    StringTokenizer st = new StringTokenizer(br.readLine());

    // input & init
    n = Integer.parseInt(st.nextToken());
    m = Integer.parseInt(st.nextToken());

    pos = new int[n+1][2];
    chk = new boolean[n+1];
    
    for(int i = 0; i <= n; i++) adj.add(new ArrayList<>());
    
    for(int i = 1; i <= n; i++){
      st = new StringTokenizer(br.readLine());
      pos[i][0] = Integer.parseInt(st.nextToken());
      pos[i][1] = Integer.parseInt(st.nextToken());
    }

    while(m-- > 0){
      st = new StringTokenizer(br.readLine());
      int u = Integer.parseInt(st.nextToken());
      int v = Integer.parseInt(st.nextToken());
      adj.get(u).add(new long[]{0, v});
      adj.get(v).add(new long[]{0, u});
    }

    // solve 
    for(int i = 1; i <= n; i++){
      for(int j = i + 1; j <= n; j++){
        long dx = pos[i][0] - pos[j][0];
        long dy = pos[i][1] - pos[j][1];
        long dist = dx*dx + dy*dy;
        adj.get(i).add(new long[]{dist, j});
        adj.get(j).add(new long[]{dist, i});
      }
    }

    chk[1] = true;
    for(long[] nxt : adj.get(1)) pq.add(new long[]{nxt[0], 1, nxt[1]});

    int cnt = 0;
    double ans = 0.;
    while(cnt < n - 1){
      long[] cur = pq.poll();
      if(chk[(int)cur[2]]) continue;
      cnt++;
      chk[(int)cur[2]] = true;
      ans += Math.sqrt(cur[0]);
      
      for(long[] nxt : adj.get((int)cur[2])){
        if(chk[(int)nxt[1]]) continue;
        pq.add(new long[]{nxt[0], cur[2], nxt[1]});
      }
    }

    // output
    System.out.println(String.format("%.2f", ans));
  }
}